specs/obmd-integration.md

```diff
@@ -0,0 +1,129 @@
+   already been made.
+
+(1) ensures that if HIL does not successfully commit, it will fail
+"safe." OBM operations will fail because the token stored in the
+database is invalid, but no access violations will occur.
+
+(2) ensures that an inconsistency between HIL and obmd can be corrected
+just by calling enable or disable. Note that this means in the event of
+a failure, HIL and obmd *may* become out of sync, but the consequences
+of this basically amount to some bad error messages.
+
+If `enable_obm` creates a token and fails to commit, this is not a problem:
+per the above, inconsistencies are easily fixable, and it cannot cause
+an access violation. While there is valid token, it is lost and not
+available to any user.
+
+# Alternative Solutions
+
+Possible variations:
+
+* We could leave the non-console functionality managed directly by HIL.
+  This was discussed in the issue and rejected.
+  * One problem is that the obm information (hosts, passwords, etc) now
+    has to be in too places.
+  * Additionally, for this to work we need direct driver support in two
+    daemons; moving it all into one place allows us to simplify the way
+    OBMs work in HIL (see the discussion about un-drivertizing OBMs in
+    HIL).
+* We could attempt to shut down the console in `project_detach_node`,
+  rather than just failing if it is connected. This would obviate
+  `enable_obm` and `disable_obm`.
+  * This was explored and turned out to be very tricky to get right; see
+    the discussion on issue #417. It is also inconsistent with our
+    approach to resource management elsewhere. @okrieg pointed this out
+    and suggested the current approach at the weekly HIL meeting on Nov
+    21.
+* We could attempt to do streaming of the console in HIL itself, rather
+  than have a separate daemon.
+  * This was tried in PR #483, but abandoned. We encountered a number of
+    difficulties:
+    * Synchronizing between request handlers was difficult. If we wanted
+      to keep HIL able to run in most WSGI environments, it meant we
+      couldn't assume the request handlers were running in the same
+      process, or even on the same machine for that matter.
+    * The solution basically mandates the use of greenlets for any kind
+      of scalability, since otherwise you're tying up a whole thread per
+      request, and these are long-lived requests. We could have just
+      documented that we expect greenlets, but it makes the api server less
+      portable.
+      * The daemon, being written in Go, can assume that request
+        handlers are light-weight.
+* We could try to proxy the console, rather than redirecting, but this
+  raises all of same the problems with doing streaming in HIL directly.
+* We could try to keep OBMs drivertized within HIL. The trade-offs have
+  been discussed above. It's not clear there are significant benefits.
+* `node_register` could just take an obmd node label (or we could just
```

Comment:
it's not clear how this will work; could you elaborate? Like, could you walk me through how the node_power_cycle call will work.

Reply:
I'm not sure I fully understand what you're confused about?

Reply:
The changes that will be made in HIL aren't clear to me.

1. If we have only one obmd server. Then would admins register the node's ipmi details directly using obmd, and then in HIL they register only the obmd node label? I am assuming HIL would know from the config file wherever the obmd is running?

2. Just spitballing here: if we want to support multiple obmd servers, then maybe we could think of obmd servers as switches in HIL. We register an obmd server with it's address and admin token. Nodes' ipmi details will be fed into the obmd and then our nodes become "ports" on the obmd server.

3. How will end users interact with obmd? Will we be changing all of HIL's ipmi related APIs (power cycle, power off etc)?

4. Database: Will the obmd have it's own database? or will it talk to the same database that HIL uses (different table perhaps?)